drv/transceivers-server/src/main.rs

```diff
@@ -0,0 +1,181 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#![no_std]
+#![no_main]
+
+use drv_sidecar_front_io::transceivers::Transceivers;
+use drv_transceivers_api::{
+    ModulesStatus, TransceiversError, NUM_PORTS, PAGE_SIZE_BYTES,
+};
+use idol_runtime::{ClientError, Leased, RequestError, R, W};
+use userlib::task_slot;
+
+task_slot!(FRONT_IO, front_io);
+
+struct ServerImpl {
+    transceivers: Transceivers,
+}
+
+impl idl::InOrderTransceiversImpl for ServerImpl {
+    fn get_modules_status(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+    ) -> Result<ModulesStatus, idol_runtime::RequestError<TransceiversError>>
+    {
+        Ok(ModulesStatus::from(
+            self.transceivers
+                .get_modules_status()
+                .map_err(TransceiversError::from)?,
+        ))
+    }
+
+    fn set_power_enable(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+        mask: u32,
+    ) -> Result<(), idol_runtime::RequestError<TransceiversError>> {
+        Ok(self
+            .transceivers
+            .set_power_enable(mask)
+            .map_err(TransceiversError::from)?)
+    }
+
+    fn clear_power_enable(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+        mask: u32,
+    ) -> Result<(), idol_runtime::RequestError<TransceiversError>> {
+        Ok(self
+            .transceivers
+            .clear_power_enable(mask)
+            .map_err(TransceiversError::from)?)
+    }
+
+    fn set_reset(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+        mask: u32,
+    ) -> Result<(), idol_runtime::RequestError<TransceiversError>> {
+        Ok(self
+            .transceivers
+            .set_reset(mask)
+            .map_err(TransceiversError::from)?)
+    }
+
+    fn clear_reset(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+        mask: u32,
+    ) -> Result<(), idol_runtime::RequestError<TransceiversError>> {
+        Ok(self
+            .transceivers
+            .clear_reset(mask)
+            .map_err(TransceiversError::from)?)
+    }
+
+    fn set_lpmode(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+        mask: u32,
+    ) -> Result<(), idol_runtime::RequestError<TransceiversError>> {
+        Ok(self
+            .transceivers
+            .set_lpmode(mask)
+            .map_err(TransceiversError::from)?)
+    }
+
+    fn clear_lpmode(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+        mask: u32,
+    ) -> Result<(), idol_runtime::RequestError<TransceiversError>> {
+        Ok(self
+            .transceivers
+            .clear_lpmode(mask)
+            .map_err(TransceiversError::from)?)
+    }
+
+    fn setup_i2c_op(
+        &mut self,
+        _msg: &userlib::RecvMessage,
+        is_read: bool,
+        reg: u8,
```

Comment:
Are there values of `reg` that are invalid?  If so, we should check and return an error.

Reply:
Not technically. `reg` represents a register address in the QSFP module, which is a 256-byte address space. I _could_ add some checking around `reg` and `num_bytes` as there is a little oddness around how the internal address register of the module works, but that could be starting to get a little cute.

Regardless, we don't expect more than a 128 byte operation here, so I will add a check for that.